spring-cloud-data-module-deployers/spring-cloud-data-module-deployer-cloudfoundry/src/main/java/org/springframework/cloud/data/module/deployer/cloudfoundry/CloudControllerTemplate.java

```diff
@@ -0,0 +1,235 @@
+		UriComponentsBuilder uriComponentsBuilder = UriComponentsBuilder.fromUri(this.endpoint)
+				.pathSegment("v2", "service_instances")
+				.queryParam("q", "space_guid:" + request.getSpaceId());
+		if (!StringUtils.isEmpty(request.getName())) {
+			uriComponentsBuilder.queryParam("q", "name:" + request.getName());
+		}
+		URI uri = uriComponentsBuilder.build().toUri();
+
+		return this.restOperations.getForObject(uri, ListServiceInstancesResponse.class);
+	}
+
+	@Override
+	public ListSharedDomainsResponse listSharedDomains(ListSharedDomainsRequest request) {
+		UriComponentsBuilder uriComponentsBuilder = UriComponentsBuilder.fromUri(this.endpoint)
+				.pathSegment("v2", "shared_domains");
+		if (!StringUtils.isEmpty(request.getName())) {
+			uriComponentsBuilder.queryParam("q", "name:" + request.getName());
+		}
+		URI uri = uriComponentsBuilder.build().toUri();
+
+		return this.restOperations.getForObject(uri, ListSharedDomainsResponse.class);
+	}
+
+	@Override
+	public ListSpacesResponse listSpaces(ListSpacesRequest request) {
+		URI uri = UriComponentsBuilder.fromUri(this.endpoint)
+				.pathSegment("v2", "spaces")
+				.queryParam("q", "name:" + request.getName())
+				.queryParam("q", "organization_guid:" + request.getOrgId())
+				.build().toUri();
+
+		return this.restOperations.getForObject(uri, ListSpacesResponse.class);
+	}
+
+	@Override
+	public RouteMappingResponse mapRoute(RouteMappingRequest request) {
+		URI uri = UriComponentsBuilder.fromUri(this.endpoint)
+				.pathSegment("v2", "routes", request.getRouteId(), "apps", request.getAppId())
+				.build().toUri();
+
+		return this.restOperations.putForObject(uri, null, RouteMappingResponse.class);
+	}
+
+	@Override
+	public RemoveServiceBindingResponse removeServiceBinding(RemoveServiceBindingRequest request) {
+		URI uri = UriComponentsBuilder.fromUri(this.endpoint)
+				.pathSegment("v2", "apps", request.getAppId(), "service_bindings", request.getBindingId())
+				.build().toUri();
+
+		this.restOperations.delete(uri);
+		return new RemoveServiceBindingResponse();
+	}
+
+	@Override
+	public UpdateApplicationResponse updateApplication(UpdateApplicationRequest request) {
+		URI uri = UriComponentsBuilder.fromUri(this.endpoint)
+				.pathSegment("v2", "apps", request.getId())
+				.build().toUri();
+
+		return this.restOperations.putForObject(uri, request, UpdateApplicationResponse.class);
+	}
+
+	@Override
+	public RouteMappingResponse unmapRoute(RouteMappingRequest request) {
+		URI uri = UriComponentsBuilder.fromUri(this.endpoint)
+				.pathSegment("v2", "routes", request.getRouteId(), "apps", request.getAppId())
+				.build().toUri();
+
+		this.restOperations.delete(uri);
+		return new RouteMappingResponse();
+	}
+
+	@Override
+	public UploadBitsResponse uploadBits(UploadBitsRequest request) {
+		URI uri = UriComponentsBuilder.fromUri(this.endpoint)
+				.pathSegment("v2", "apps", request.getId(), "bits")
+				.queryParam("async", false)
```

Comment:
REST API specifies `async` default value is `false`. Hence, this may not be needed.